keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -0,0 +1,511 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+Kubernetes relies on resource data being actively re-written to perform certain maintenance actives related to at rest storage.  Two prominent examples are the versioned schema of stored resources (i.e. the preferred storage schema changing from `v1beta1` to `v1` for a given resource) and encryption at rest (i.e. rewriting `stale` data based on a change in how the data should be encrypted).
+
+The simplest way to rewrite data is to issue no-op `update` requests via `kubetctl get <resource> | kubectl replace -`.  This is problematic for any resource that can contain a large amount of data such as Kubernetes `secrets` and it is also impractical to perform without automation as the number of resources that need migration is always growing.
+
+Unlike most Kuberentes API interactions, during storage migration, it is safe to ignore conflicts during `update` requests and it is safe to use inconsistent continue tokens during paginated `list` operations (since we only need data to be re-written, we do not care how it is re-written).
+
+This KEP aims to make it easy for users to perform storage migrations without having to worry about the details.
+
+## Motivation
+
+### Goals
+
+- Make it easy for an admin to trigger a storage migration without having to manually run `kubectl get | kubectl replace`
+- Enable infrastructure providers to use internal details of their environment to trigger storage migration (ex: triggering storage migration of Kubernetes secrets when the KMS v2 key ID has changed)
+- Make it easy for existing users of SVM to migrate to the in-tree controller
+
+### Non-Goals
+
+- Any automatic or direct integration with KMS v2
+- Any modification regarding `StorageVersion` API for HA API servers
+- Adding logic that relies on the hashed storage versions exposed via the discovery API
+
+### UNCLEAR Goals and/or Non-Goals
+
+- Automatic storage version migration for CRDs
+- Make it easy for Kuberentes developers to drop old API schemas by guaranteeing that storage migration is run automatically on SV hash changes (should this also be on a timer or API server identity?)
+- Automated Storage Version Migration via the hash exposed by the `StorageVersion` API
```

Comment:
Right we all agree that we need this to happen automatically, but we have tried to keep this KEP small so that we don't stall like the previous efforts.